### ISSUE TITLE
Turn bin/lib into subcrates within workspace?

### ISSUE THREAD

**Issue by Kixunil** [rust-lang/rust-bindgen#1970]
Bindgen is both library and binary but binary requires certain features which are not needed in build scripts (mainly `clap`). This is handled by features but because of the way features work in Cargo they are turned on by default. This is a footgun for every library author who wishes to use build script as they may forget to turn them on. An [example of this case](https://github.com/rust-rocksdb/rust-rocksdb/pull/491) was found in `rocksb`.

Spliting up library and binary fixes the issue without causing ridiculous compile errors because of missing features.
Alternative: wait eternity until https://github.com/rust-lang/rfcs/pull/3020 is implemented.

**Comment by emilio** [comment]
Yeah, this is a reasonable idea, but I'm not sure if it's possible without breaking either binary or library users unfortunately, as both crates would have to have the same name...

**Comment by Kixunil** [comment]
Is your concern that people will run `cargo install bindgen` and it'll not work? If yes, is this something widely used in automated setups?

**Comment by emilio** [comment]
Yes, and whether it's widely used, I'm not sure.

**Comment by Kixunil** [comment]
Perhaps a way to solve it gracefully is to create a new crate called `bindgen-bin` which would be a copy of `[bin]` from `bindgen`. Once it's ready add an unconditional warning to the binary in `bindgen` saying it's deprecated and users need to switch to `bindgen-bin`. Then remove it after 6 months or something.

Do you think that could work?

**Comment by emilio** [comment]
That might be reasonable (sorry for the lag, exams and work :/)

**Comment by emilio** [comment]
Done in #2284
